### PR TITLE
[Schema Registry] Adding Support for JSON and Custom Schema Formats

### PR DESCRIPTION
**Pull request by m-redding**
These changes add a new Schema Format option for both JSON and Custom serialization.

# Contributing to the Azure SDK

Please see our [CONTRIBUTING.md](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md) if you are not familiar with contributing to this repository or have questions.

For specific information about pull request etiquette and best practices, see [this section](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#pull-request-etiquette-and-best-practices).

### CODE REVIEW

File: sdk/schemaregistry/Azure.Data.SchemaRegistry/src/SchemaFormat.cs
Comment (m-redding) [inline sdk/schemaregistry/Azure.Data.SchemaRegistry/src/SchemaFormat.cs:23]: Not sure about the capitalizations here, should it be JSONValue="JSON" or JsonValue="Json" instead?

Comment (jsquire) [reply sdk/schemaregistry/Azure.Data.SchemaRegistry/src/SchemaFormat.cs:23]: I think `JSON` make sense here; that's the [official capitalization](https://www.json.org/json-en.html).

**Review by jsquire — COMMENTED** [review]
LGTM;  I'll defer to Josh for approval.

File: sdk/schemaregistry/Azure.Data.SchemaRegistry/src/SchemaRegistryClient.cs
Comment (JoshLove-msft) [inline sdk/schemaregistry/Azure.Data.SchemaRegistry/src/SchemaRegistryClient.cs:281]: nit: extract into method

File: sdk/schemaregistry/Azure.Data.SchemaRegistry/tests/SchemaRegistryClientLiveTests.cs
Comment (JoshLove-msft) [inline sdk/schemaregistry/Azure.Data.SchemaRegistry/tests/SchemaRegistryClientLiveTests.cs:50]: nit: consider adding TestCases to each of the existing tests and passing in the format as a parameter. I think this would make us re-record all tests because the test names will change.

**Comment by JoshLove-msft** [comment]
LGTM - let's hold off on merging until we ship the new GA version.

File: sdk/schemaregistry/Azure.Data.SchemaRegistry/api/Azure.Data.SchemaRegistry.netstandard2.0.cs
Comment (JoshLove-msft) [inline sdk/schemaregistry/Azure.Data.SchemaRegistry/api/Azure.Data.SchemaRegistry.netstandard2.0.cs:10]: I think we need some tests for Custom. What is the content type for custom?

File: sdk/schemaregistry/Azure.Data.SchemaRegistry/src/SchemaRegistryClient.cs
Comment (JoshLove-msft) [inline sdk/schemaregistry/Azure.Data.SchemaRegistry/src/SchemaRegistryClient.cs:6]: nit: remove

File: sdk/schemaregistry/Azure.Data.SchemaRegistry/src/SchemaFormat.cs
Comment (JoshLove-msft) [inline sdk/schemaregistry/Azure.Data.SchemaRegistry/src/SchemaFormat.cs:29]: Sounds like we should use JSON schema content type here.

Comment (m-redding) [reply sdk/schemaregistry/Azure.Data.SchemaRegistry/src/SchemaFormat.cs:29]: Not sure if this update is exactly what you meant, so let me know if I need to adjust it

File: sdk/schemaregistry/Azure.Data.SchemaRegistry/api/Azure.Data.SchemaRegistry.netstandard2.0.cs
Comment (m-redding) [reply sdk/schemaregistry/Azure.Data.SchemaRegistry/api/Azure.Data.SchemaRegistry.netstandard2.0.cs:10]: Added the stubs for them but it seems like there isn't a true consensus on what the API will look like for custom, so I will add those in once the new swagger is in

File: sdk/schemaregistry/Azure.Data.SchemaRegistry/tests/SchemaRegistryClientLiveTests.cs
Comment (m-redding) [reply sdk/schemaregistry/Azure.Data.SchemaRegistry/tests/SchemaRegistryClientLiveTests.cs:50]: I just recorded the json test cases for now, the Avro test cases throw "Schema of type Avro cannot be created within group". I'm not sure if this is something that will be fixed once the json changes are all in on the service side, otherwise I may need to add additional test resources.

File: sdk/schemaregistry/Azure.Data.SchemaRegistry/src/Generated/Models/ContentType.cs
Comment (JoshLove-msft) [inline sdk/schemaregistry/Azure.Data.SchemaRegistry/src/Generated/Models/ContentType.cs:11]: We will need to make this not public.

File: sdk/schemaregistry/Azure.Data.SchemaRegistry/src/Generated/SchemaRestClient.cs
Comment (JoshLove-msft) [inline sdk/schemaregistry/Azure.Data.SchemaRegistry/src/Generated/SchemaRestClient.cs:372]: We probably need to come up with some conversion method to go from a SchemaFormat to a ContentType.

File: sdk/schemaregistry/Azure.Data.SchemaRegistry/src/Generated/Models/ContentType.cs
Comment (JoshLove-msft) [reply sdk/schemaregistry/Azure.Data.SchemaRegistry/src/Generated/Models/ContentType.cs]: See https://github.com/Azure/autorest.csharp#make-a-model-internal

File: sdk/schemaregistry/Azure.Data.SchemaRegistry/src/Generated/SchemaRestClient.cs
Comment (m-redding) [inline sdk/schemaregistry/Azure.Data.SchemaRegistry/src/Generated/SchemaRestClient.cs:258]: TODO: which response to return

File: sdk/schemaregistry/Azure.Data.SchemaRegistry/src/Generated/SchemaRestClient.cs
Comment (m-redding) [inline sdk/schemaregistry/Azure.Data.SchemaRegistry/src/Generated/SchemaRestClient.cs:297]: TODO: again which response

File: sdk/schemaregistry/Azure.Data.SchemaRegistry/src/Generated/SchemaRestClient.cs
Comment (m-redding) [inline sdk/schemaregistry/Azure.Data.SchemaRegistry/src/Generated/SchemaRestClient.cs:276]: TODO: is this change from Stream to object a problem?

File: sdk/schemaregistry/Azure.Data.SchemaRegistry/src/Generated/SchemaRestClient.cs
Comment (JoshLove-msft) [inline sdk/schemaregistry/Azure.Data.SchemaRegistry/src/Generated/SchemaRestClient.cs:237]: Not sure why this changed.

File: sdk/schemaregistry/Azure.Data.SchemaRegistry/api/Azure.Data.SchemaRegistry.netstandard2.0.cs
Comment (JoshLove-msft) [inline sdk/schemaregistry/Azure.Data.SchemaRegistry/api/Azure.Data.SchemaRegistry.netstandard2.0.cs:8]: Hmm where is this coming from? I don't see it in the generated code.

File: sdk/schemaregistry/Azure.Data.SchemaRegistry/src/autorest.md
Comment (JoshLove-msft) [inline sdk/schemaregistry/Azure.Data.SchemaRegistry/src/autorest.md:20]:
```suggestion
    $.produces = ["application/json; serialization=Avro", "application/json; serialization=json", "application/octet-stream"]
```

File: sdk/schemaregistry/Azure.Data.SchemaRegistry/src/autorest.md
Comment (JoshLove-msft) [inline sdk/schemaregistry/Azure.Data.SchemaRegistry/src/autorest.md:48]:
```suggestion
        "application/json; serialization=json",
```

File: sdk/schemaregistry/Azure.Data.SchemaRegistry/src/autorest.md
Comment (JoshLove-msft) [inline sdk/schemaregistry/Azure.Data.SchemaRegistry/src/autorest.md:97]:
```suggestion
        "application/json; serialization=json",
```

File: sdk/schemaregistry/Azure.Data.SchemaRegistry/src/autorest.md
Comment (JoshLove-msft) [inline sdk/schemaregistry/Azure.Data.SchemaRegistry/src/autorest.md:20]: nit: put each element of array on a separate line like in the other transforms.

File: sdk/schemaregistry/Azure.Data.SchemaRegistry/api/Azure.Data.SchemaRegistry.netstandard2.0.cs
Comment (m-redding) [reply sdk/schemaregistry/Azure.Data.SchemaRegistry/api/Azure.Data.SchemaRegistry.netstandard2.0.cs]: I forgot to re-export the API, it's gone now that I have

File: sdk/schemaregistry/Azure.Data.SchemaRegistry/tests/SessionRecords/SchemaRegistryClientLiveTests/CanGetSchema(%Json%)Async.json
Comment (JoshLove-msft) [inline sdk/schemaregistry/Azure.Data.SchemaRegistry/tests/SessionRecords/SchemaRegistryClientLiveTests/CanGetSchema(%Json%)Async.json:4]: This test should be failing now in live mode. We will need to introduce a new api version in SchemaRegistryClientOptions.

File: sdk/schemaregistry/Azure.Data.SchemaRegistry/src/SchemaFormat.cs
Comment (JoshLove-msft) [inline sdk/schemaregistry/Azure.Data.SchemaRegistry/src/SchemaFormat.cs:14]: Why did we need to add these?

File: sdk/schemaregistry/Azure.Data.SchemaRegistry/tests/SessionRecords/SchemaRegistryClientLiveTests/CanGetSchema(%Json%)Async.json
Comment (m-redding) [reply sdk/schemaregistry/Azure.Data.SchemaRegistry/tests/SessionRecords/SchemaRegistryClientLiveTests/CanGetSchema(%Json%)Async.json:4]: Yeah the live tests are all failing at the moment so I'm working on getting those working

File: sdk/schemaregistry/Azure.Data.SchemaRegistry/src/SchemaFormat.cs
Comment (m-redding) [reply sdk/schemaregistry/Azure.Data.SchemaRegistry/src/SchemaFormat.cs:14]: I added these because the generator was setting AvroValue as "application/json; serialization=Avro" (etc) and it was breaking the SchemaFormat Tests, to me it seemed like the value should be "Avro"? But if it's not breaking/not an issue, I can switch them back

Comment (JoshLove-msft) [reply sdk/schemaregistry/Azure.Data.SchemaRegistry/src/SchemaFormat.cs:14]: Ah I see. I didn't remember that the ToString used these values. You are right we need to add these back.

File: sdk/schemaregistry/Azure.Data.SchemaRegistry/src/autorest.md
Comment (JoshLove-msft) [inline sdk/schemaregistry/Azure.Data.SchemaRegistry/src/autorest.md:54]: As you discovered, we can't do this because it would break the ToString behavior.

**Comment by m-redding** [comment]
Working on re-recording the tests

File: sdk/schemaregistry/Azure.Data.SchemaRegistry/src/SchemaFormat.cs
Comment (JoshLove-msft) [reply sdk/schemaregistry/Azure.Data.SchemaRegistry/src/SchemaFormat.cs:65]: We could also rename SchemaFormatInternal to ContentTypeInternal.

File: sdk/schemaregistry/Azure.Data.SchemaRegistry/src/ContentTypeInternal.cs
Comment (JoshLove-msft) [inline sdk/schemaregistry/Azure.Data.SchemaRegistry/src/ContentTypeInternal.cs:11]: I'm not sure that we need this customization. Could we just update the swagger to have the x-ms-enum name that we want - probably just ContentType would be fine.

File: sdk/schemaregistry/Azure.Data.SchemaRegistry/src/SchemaFormat.cs
Comment (JoshLove-msft) [inline sdk/schemaregistry/Azure.Data.SchemaRegistry/src/SchemaFormat.cs:23]: nit: I'd suggest leaving the ContentType property off of here and just add an internal method, ToContentType that returns the correct ContentType.

File: sdk/schemaregistry/Azure.Data.SchemaRegistry/tests/SchemaRegistryClientLiveTests.cs
Comment (JoshLove-msft) [inline sdk/schemaregistry/Azure.Data.SchemaRegistry/tests/SchemaRegistryClientLiveTests.cs:16]: nit: remove the mode

File: sdk/schemaregistry/Azure.Data.SchemaRegistry/tests/SessionRecords/SchemaRegistryClientLiveTests/CanGetSchema(%Json%).json
Comment (JoshLove-msft) [inline sdk/schemaregistry/Azure.Data.SchemaRegistry/tests/SessionRecords/SchemaRegistryClientLiveTests/CanGetSchema(%Json%).json:1]: Did we need to re-record these because of the parameter changing the name? 
Are there extra recordings that need to be deleted?

File: sdk/schemaregistry/Azure.Data.SchemaRegistry/api/Azure.Data.SchemaRegistry.netstandard2.0.cs
Comment (conniey) [inline sdk/schemaregistry/Azure.Data.SchemaRegistry/api/Azure.Data.SchemaRegistry.netstandard2.0.cs:7]: Not sure if you meant to, but there is an empty sc.txt file.

Comment (m-redding) [reply sdk/schemaregistry/Azure.Data.SchemaRegistry/api/Azure.Data.SchemaRegistry.netstandard2.0.cs:7]: That got checked in accidentally!

**Comment by JoshLove-msft** [comment]
Looks like we will need to re-record tests because of new API version.

File: sdk/schemaregistry/Azure.Data.SchemaRegistry/tests/SessionRecords/SchemaRegistryClientLiveTests/CanGetSchema(%Json%).json
Comment (m-redding) [reply sdk/schemaregistry/Azure.Data.SchemaRegistry/tests/SessionRecords/SchemaRegistryClientLiveTests/CanGetSchema(%Json%).json:1]: Since the API changed that means that all of them need to be re-recorded?

**Review by JoshLove-msft — APPROVED** [review]

**Pull request merged**
